packages/js/src/dashboard/services/data-tracker.js

```diff
@@ -0,0 +1,88 @@
+import { noop } from "lodash";
+
+/**
+ * @typedef {Object} SetupStepsTrackingData The setup steps tracking data.
+ * @property {string} [setupWidgetLoaded] Whether Site Kit setup widget is loaded.
+ * @property {string} [firstInteractionStage] The first stage of the Site Kit setup widget the user interacted with.
+ * @property {string} [lastInteractionStage] The last stage of the Site Kit setup widget the user interacted with.
+ * @property {string} [setupWidgetTemporarilyDismissed] Stores if the Site Kit setup widget has been temporarily dismissed.
+ * @property {string} [setupWidgetPermanentlyDismissed] Stores if the Site Kit setup widget has been temporarily dismissed.
+ */
+
+/**
+ * Represents data that we want to track.
+ */
+export class DataTracker {
+	#setupStepsTracking;
+	#dataProvider;
+	#remoteDataProvider;
+
+	/**
+	 * @param {SetupStepsTrackingData} setupStepsTrackingData The setup steps tracking data.
+	 * @param {import("./data-provider").DataProvider} dataProvider The data provider.
+	 * @param {import("./remote-data-provider").RemoteDataProvider} remoteDataProvider The remote data provider.
+	 */
+	constructor( setupStepsTrackingData, dataProvider, remoteDataProvider ) {
```

Comment:
`setupStepsTrackingData` is a very specific name. I would make it more general to allow future tracking of unrelated to site kit.

Reply:
See the comment about the `DataTracker` architecture.